### PR TITLE
GHA automatically  build & test tvOS integration tests on tvOS Simulator

### PR DESCRIPTION
**Pull request by sunmou99**
Fixed simulator test failures.
Enabled the following tvOS simulator tests: auth, database, functions, remote_config, storage, firestore
The following will **not** be built nor tested: admob, analytics, dynamic_links, messaging, installations

Test on your mac:
1. Install prerequisites:
build_scripts/tvos/install_prereqs.sh
pip install -r scripts/gha/requirements.txt
python scripts/gha/restore_secrets.py --passphrase ${passphrase}
2. Build tvos integration tests (View code changes in this PR):
python scripts/gha/build_testapps.py --t auth,database,firestore,functions,remote_config,storage --p tvOS --output_directory ${output_directory}
3. Test tvos integration tests (View change in [PR](https://github.com/firebase/firebase-cpp-sdk/pull/489)):
python scripts/gha/test_simulator.py --testapp_dir ${output_directory}/testapps-local-build
4. View test summary at ${output_directory}/testapps-local-build/test-results-simulator-test.log

### CODE REVIEW

File: .github/workflows/integration_tests.yml [hunk 1]
```diff
@@ -9,8 +9,8 @@ on:
   workflow_dispatch:
     inputs:
       platforms:
-        description: 'CSV of Desktop, Android and/or iOS'
-        default: 'Desktop,Android,iOS'
+        description: 'CSV of Desktop, Android and/or iOS, tvOS'
```
Comment (jonsimantov) [inline .github/workflows/integration_tests.yml:12]: nit: "CSV of Desktop, Android, iOS, and/or tvOS

File: scripts/gha/build_testapps.py [hunk 2]
```diff
@@ -558,24 +571,80 @@ def _build_ios(
...
-            ios_sdk=_IOS_SDK_DEVICE,
+            apple_platfrom=_IOS,
+            apple_sdk=_IOS_SDK_DEVICE,
             configuration="Debug"))
 
     xcodebuild.generate_unsigned_ipa(
         output_dir=build_dir, configuration="Debug")
 
 
+def _build_tvos(
```
Comment (jonsimantov) [inline scripts/gha/build_testapps.py:594]: Is this different enough from iOS to require a separate function? Or could it just be a parameter to the ios one?

Comment (sunmou99) [reply scripts/gha/build_testapps.py:594]: The main differences are: 1) Currently, tvOS won't build .ipa for real device. 2) iOS entitlements introduce errors in tvOS apps.
But, yeah, we could add a few more parameters to reuse the iOS one.

File: .github/workflows/integration_tests.yml [hunk 3]
```diff
@@ -581,6 +583,103 @@ jobs:
...
+    if: contains(needs.prepare_matrix.outputs.matrix_platform, 'tvOS') && contains(needs.prepare_matrix.outputs.matrix_os, 'macos-latest') && needs.prepare_matrix.outputs.apis != '' && !cancelled() && !failure()
+    strategy:
+      fail-fast: false
+    steps:
+      - uses: actions/checkout@v2
+        with:
+          ref: ${{needs.prepare_matrix.outputs.github_ref}}
+          submodules: true
+      - name: Set env vars (macOS)
+        run: echo "VCPKG_TRIPLET=x64-osx" >> $GITHUB_ENV
```
Comment (vimanyu) [inline .github/workflows/integration_tests.yml:599]: Don't think we need vcpkg for ios or tvos.

File: .github/workflows/integration_tests.yml [hunk 4]
```diff
@@ -581,6 +583,103 @@ jobs:
...
+      - name: Set env vars (all)
+        shell: bash
+        run: echo "VCPKG_RESPONSE_FILE=external/vcpkg_${{ env.VCPKG_TRIPLET }}_response_file.txt" >> $GITHUB_ENV
```
Comment (vimanyu) [inline .github/workflows/integration_tests.yml:602]: same as above.

File: scripts/gha/build_testapps.py [hunk 5]
```diff
@@ -326,6 +340,21 @@ def _build(
           Failure(testapp=testapp, platform=_IOS, error_message=str(e)))
     logging.info("END %s, %s", testapp, _IOS)
 
+  if _TVOS in platforms and api_config.tvos_target != "":
```
Comment (vimanyu) [inline scripts/gha/build_testapps.py:343]: nit: more readable
`if _TVOS in platforms and api_config.tvos_target`

File: scripts/gha/build_testapps.py [hunk 6]
```diff
@@ -558,24 +571,80 @@ def _build_ios(
...
 
     xcodebuild.generate_unsigned_ipa(
         output_dir=build_dir, configuration="Debug")
 
 
+def _build_tvos(
+    sdk_dir, xcframework_exist, project_dir, repo_dir, api_config, tvos_sdk):
+  """Builds an iOS application (.app, .ipa or both)."""
+  if not xcframework_exist:
+    _build_xcframework_from_repo(repo_dir, "tvos", "tvos_build", api_config)
```
Comment (vimanyu) [inline scripts/gha/build_testapps.py:598]: Should we use a variable for the build directory `tvos_build` as it is being hardcoded in many places?

File: scripts/gha/summarize_test_results.py [hunk 7]
```diff
@@ -317,12 +318,12 @@ def combine_configs(configs):
 #     ['ubuntu', 'windows', 'macos']
 #     -> ['All os']
 def combine_config(config, platform, k):
-  if k == 1 and (platform == "android" or platform == "ios"):
+  if k == 1 and (platform == "android" or platform == "ios" or platform == "tvos"):
```
Comment (vimanyu) [inline scripts/gha/summarize_test_results.py:321]: `platform in ("android", "ios", "tvos")`

File: scripts/gha/build_testapps.py (hunk 2 above)
Comment (vimanyu) [reply scripts/gha/build_testapps.py:594]: Interesting. Do we know why the entitlement file is not working on tvos? auth, messaging and dynamic links are using entitlement files on iOS. Am surprised auth is working without it on tvOS. We do not care about dynamic links but messaging might need it once we have it up and running on tvos.

**Review by vimanyu — COMMENTED** [review]
Thanks for taking care of this! 
Some minor comments in code and some other notes,
- "messaging" integration test had some errors when I ran it locally on xcode. messaging is not ready for tvos yet and we should probably turn off the integration tests for it until it is ready.
-  I see these non-blocking errors on building testapps locally on my mac,
```
I0630 18:23:46.513381 4530540032 build_testapps.py:356] END storage, tvOS
I0630 18:23:46.513784 4530540032 build_testapps.py:270] END building for storage
I0630 18:23:46.956192 4530540032 build_testapps.py:675] Deleting <firebase_cpp_dir>/tvos_test/testapps-...
W0630 18:23:46.956428 4530540032 build_testapps.py:682] Failed to remove directory:
[Errno 2] No such file or directory: '<firebase_checkout>/tvos_test/testapps-'
I0630 18:23:47.354665 4530540032 build_testapps.py:412] BUILD SUMMARY:
```
Also, I am guessing we are missing a part for the directory name as it ends with a hyphen.

- Got this error while trying to run the "built" tests
```
I0630 18:33:19.129165 4637900288 test_simulator.py:391] Download simulator: xcversion simulators --install=tvOS 14.0
Traceback (most recent call last):
  File "<firebase_cpp_dir>/scripts/gha/test_simulator.py", line 611, in <module>
    app.run(main)
  File "/Users/<user>/Library/Python/3.9/lib/python/site-packages/absl/app.py", line 303, in run
    _run_main(main, args)
  File "/Users/<user>/Library/Python/3.9/lib/python/site-packages/absl/app.py", line 251, in _run_main
    sys.exit(main(argv))
  File "<firbease_dir>/scripts/gha/test_simulator.py", line 231, in main
    device_id = _create_and_boot_simulator("tvOS", device_name, device_os)
  File "/Users/<user>/dev/cpp/firebase/firebase-cpp-sdk/scripts/gha/test_simulator.py", line 392, in _create_and_boot_simulator
    subprocess.run(args=args, check=False)
  File "/Users/<user>/homebrew/Cellar/python@3.9/3.9.5/Frameworks/Python.framework/Versions/3.9/lib/python3.9/subprocess.py", line 505, in run
    with Popen(*popenargs, **kwargs) as process:
  File "/Users/<user>/homebrew/Cellar/python@3.9/3.9.5/Frameworks/Python.framework/Versions/3.9/lib/python3.9/subprocess.py", line 951, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "/Users/<user>/homebrew/Cellar/python@3.9/3.9.5/Frameworks/Python.framework/Versions/3.9/lib/python3.9/subprocess.py", line 1821, in _execute_child
    raise child_exception_type(errno_num, err_msg, err_filename)
FileNotFoundError: [Errno 2] No such file or directory: 'xcversion'
```

**Comment by sunmou99** [comment]
Thanks @vimanyu 

- Messaging is disabled. 
- Non-blocking errors are not errors, I will change the log level.
- Yeah, you're right, there is a "missing a part" for the directory name, but it only exists in GHA. I'll update the name to avoid confusion.
- Install `xcversion` should solve the problem. https://github.com/xcpretty/xcode-install#xcodeinstall

File: scripts/gha/build_testapps.py (hunk 6 above)
Comment (sunmou99) [reply scripts/gha/build_testapps.py]: Good point!

File: scripts/gha/build_testapps.py (hunk 2 above)
Comment (sunmou99) [reply scripts/gha/build_testapps.py]: Reused the same function with more parameters.
It seems entitlements do not affect current tests in our workflow.
But, we still need to handle this entitlements stuff later. (Probably in other PR.)

Comment (vimanyu) [reply scripts/gha/build_testapps.py]: Sounds good. Maybe we can leave a TODO in code and take out the commented out code for entitlements for tvos?

**Comment by vimanyu** [comment]
> Thanks @vimanyu
> 
> * Messaging is disabled.
> * Non-blocking errors are not errors, I will change the log level.
> * Yeah, you're right, there is a "missing a part" for the directory name, but it only exists in GHA. I'll update the name to avoid confusion.
> * Install `xcversion` should solve the problem. https://github.com/xcpretty/xcode-install#xcodeinstall

Thanks for the clarification! Confirmed that I could successfully run all the tests (including firestore) locally on the simulators via these scripts. Let us sync up before merging this so we can co-ordinate changes in other workflows.

**Review by vimanyu — APPROVED** [review]
LGTM!

**Pull request merged**